### PR TITLE
Pinecone update to Readme and docker compose for ChatQnA

### PR DESCRIPTION
**Pull request by pallavijaini0525**
## Description

Added the Readme and Docker compose files for Pinecone DB for ChatQnA usecase

## Issues

https://github.com/opea-project/GenAIExamples/issues/411

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Others (enhancement, documentation, validation, etc.)

### CODE REVIEW

**Comment by chensuyue** [comment]
1. Do we need a separate readme?   
2. `ChatQnA/docker/xeon/docker_compose_pinecone.yaml` --> `ChatQnA/docker/xeon/compose_pinecone.yaml`
3. Please add an example e2e test, like this one, https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/tests/test_chatqna_on_xeon.sh

**Comment by pallavijaini0525** [comment]
@chensuyue -> Sure i will add the tests, can you please help in adding the PINECONE_KEY to the project variables simialr to HF_TOKEN, so i can use that key for the tests.

@ZePan110 has added for GenAIComps

**Comment by chensuyue** [comment]
> @chensuyue -> Sure i will add the tests, can you please help in adding the PINECONE_KEY to the project variables simialr to HF_TOKEN, so i can use that key for the tests.
> 
> @ZePan110 has added for GenAIComps

https://github.com/opea-project/GenAIExamples/pull/579

**Comment by pallavijaini0525** [comment]
Tests are added and file is renamed to compose_pinecone.yaml

File: ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh [hunk 1]
```diff
@@ -0,0 +1,232 @@
+#!/bin/bash
```
Comment (chensuyue) [inline ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh:1]: '_' before test means a test is not ready for test, so please remove it from the script name.

Comment (chensuyue) [reply ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh:1]: And we refine the test structure recent days, please follow this one to make the changes, https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/tests/test_chatqna_on_xeon.sh.

**Comment by arun-gupta** [comment]
What is the current status of this merge?

In addition, can `opea/*-pinecone` Docker images be pushed to Docker Hub? This will simplify the overall developer experience.

**Comment by pallavijaini0525** [comment]
Only the docker compose and Readme files are not merged which are part of this PR, In GenAIComps, pinecone related changes are merged

**Comment by arun-gupta** [comment]
Seems like @chensuyue requested some changes. Can we make them so that it can be integrated?

File: ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh (hunk 1 above)
Comment (pallavijaini0525) [reply ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh:1]: updated the tests as per the above suggestion

**Comment by pallavijaini0525** [comment]
@chensuyue  -> All issues are resolved, please review

File: ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
```
Comment (chensuyue) [inline ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh:3]: In the new test structure, the test scripts name should be `ChatQnA/tests/test_compose_pinecone_on_xeon.sh`

File: ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+WORKPATH=$(dirname "$PWD")
+LOG_PATH="$WORKPATH/tests"
+ip_address=$(hostname -I | awk '{print $1}')
+
+function build_docker_images() {
+    cd $WORKPATH
+    git clone https://github.com/opea-project/GenAIComps.git
+    cd GenAIComps
+
+    docker build -t opea/embedding-tei:latest -f comps/embeddings/langchain/docker/Dockerfile .
```
Comment (chensuyue) [inline ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh:17]: https://github.com/opea-project/GenAIExamples/blob/558ea3bb7fc915a0622aaf13c76cf52303fee164/ChatQnA/tests/test_compose_on_xeon.sh#L23, use compose build to build images. So we can define the repo and tag for the test image. This is used by release test.    
IMAGE_REPO=${IMAGE_REPO:-"opea"}
IMAGE_TAG=${IMAGE_TAG:-"latest"}

File: ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh (hunk 1 above)
Comment (chensuyue) [reply ChatQnA/tests/_test_chatqna_pinecone_on_xeon.sh:1]: Why duplicate the test scripts, you should remove this one.

File: ChatQnA/tests/test_chatqna_pinecone_on_xeon.sh [hunk 4]
```diff
@@ -0,0 +1,258 @@
+#!/bin/bash
```
Comment (chensuyue) [inline ChatQnA/tests/test_chatqna_pinecone_on_xeon.sh:1]: In the new test structure, the test scripts name should be `ChatQnA/tests/test_compose_pinecone_on_xeon.sh`

File: ChatQnA/tests/test_chatqna_pinecone_on_xeon.sh [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    docker compose -f build.yaml build ${service_list} --no-cache > ${LOG_PATH}/docker_image_build.log
+
+    docker pull ghcr.io/huggingface/tgi-gaudi:2.0.1
+    docker pull ghcr.io/huggingface/text-embeddings-inference:cpu-1.5
+
+    docker images && sleep 1s
+}
+
+function start_services() {
+    cd $WORKPATH/docker/xeon
```
Comment (chensuyue) [inline ChatQnA/tests/test_chatqna_pinecone_on_xeon.sh:32]: Wrong test path.

**Comment by chensuyue** [comment]
> @chensuyue -> All issues are resolved, please review

Because the test scripts name not adapt to the new structure, the test is actually not triggered.

File: ChatQnA/tests/test_compose_pinecone_on_xeon.sh [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+set -e
+IMAGE_REPO=${IMAGE_REPO:-"opea"}
+IMAGE_TAG=${IMAGE_TAG:-"latest"}
+echo "REGISTRY=IMAGE_REPO=${IMAGE_REPO}"
+echo "TAG=IMAGE_TAG=${IMAGE_TAG}"
+export REGISTRY=${IMAGE_REPO}
+export TAG=latest
```
Comment (chensuyue) [inline ChatQnA/tests/test_compose_pinecone_on_xeon.sh:11]: `export TAG=latest` -> `export TAG=${IMAGE_TAG}`

**Comment by chensuyue** [comment]
Seems the test failed in this part, please check, 
image

**Comment by pallavijaini0525** [comment]
All tests are validated in IDC and it is working as expected. llm-service is taking longer time to respond, in my IDC machine, it took around 3 minutes. Attaching the screenshot of successful tests.

Dataprep, insertion, deletion, retrieval, llm service, all are working as expected.

@arun-gupta  @chensuyue 

image
image

**Comment by chensuyue** [comment]
I know the issue. In CI test, we use the image like this, we don't use latest tag. 
image  
But in your yaml, the tag has been froze, so the env parameter `TAG` are not read. please check my comments to fix the test.

File: ChatQnA/benchmark/performance/helm_charts/customize.yaml [hunk 7]
```diff
@@ -6,7 +6,7 @@ HUGGINGFACEHUB_API_TOKEN: ${HF_TOKEN}
 podSpecs:
   - name: chatqna-backend-server-deploy
     spec:
-      image_name: opea/chatqna-no-wrapper
+      image_name: opea/chatqna
```
Comment (chensuyue) [inline ChatQnA/benchmark/performance/helm_charts/customize.yaml:9]: Those change under `benchmark` shouldn't include in your PR, please fix the rebasing.

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 8]
```diff
@@ -0,0 +1,201 @@
+
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+version: "3.8"
+
+services:
+  dataprep-pinecone-service:
+    image: opea/dataprep-pinecone:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:9]: ${REGISTRY:-opea}/dataprep-redis:${TAG:-latest}

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    volumes:
+      - "./data:/data"
+    shm_size: 1g
+    environment:
+      no_proxy: ${no_proxy}
+      http_proxy: ${http_proxy}
+      https_proxy: ${https_proxy}
+    command: --model-id ${EMBEDDING_MODEL_ID} --auto-truncate
+  embedding:
+    image: opea/embedding-tei:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:40]: same update here.

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+      no_proxy: ${no_proxy}
+      http_proxy: ${http_proxy}
+      https_proxy: ${https_proxy}
+      TEI_EMBEDDING_ENDPOINT: ${TEI_EMBEDDING_ENDPOINT}
+      LANGCHAIN_API_KEY: ${LANGCHAIN_API_KEY}
+      LANGCHAIN_TRACING_V2: ${LANGCHAIN_TRACING_V2}
+      LANGCHAIN_PROJECT: "opea-embedding-service"
+    restart: unless-stopped
+  retriever:
+    image: opea/retriever-pinecone:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:57]: same update here.

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    environment:
+      no_proxy: ${no_proxy}
+      http_proxy: ${http_proxy}
+      https_proxy: ${https_proxy}
+      HUGGINGFACEHUB_API_TOKEN: ${HUGGINGFACEHUB_API_TOKEN}
+      HF_HUB_DISABLE_PROGRESS_BARS: 1
+      HF_HUB_ENABLE_HF_TRANSFER: 0
+    command: --model-id ${RERANK_MODEL_ID} --auto-truncate
+  reranking:
+    image: opea/reranking-tei:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:89]: same update here.

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    environment:
+      no_proxy: ${no_proxy}
+      http_proxy: ${http_proxy}
+      https_proxy: ${https_proxy}
+      HF_TOKEN: ${HUGGINGFACEHUB_API_TOKEN}
+      HF_HUB_DISABLE_PROGRESS_BARS: 1
+      HF_HUB_ENABLE_HF_TRANSFER: 0
+    command: --model-id ${LLM_MODEL_ID}
+  llm:
+    image: opea/llm-tgi:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:125]: ****

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+      TGI_LLM_ENDPOINT: ${TGI_LLM_ENDPOINT}
+      HUGGINGFACEHUB_API_TOKEN: ${HUGGINGFACEHUB_API_TOKEN}
+      HF_HUB_DISABLE_PROGRESS_BARS: 1
+      HF_HUB_ENABLE_HF_TRANSFER: 0
+      LANGCHAIN_API_KEY: ${LANGCHAIN_API_KEY}
+      LANGCHAIN_TRACING_V2: ${LANGCHAIN_TRACING_V2}
+      LANGCHAIN_PROJECT: "opea-llm-service"
+    restart: unless-stopped
+  chaqna-xeon-backend-server:
+    image: opea/chatqna:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:145]: same update here.

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+      - http_proxy=${http_proxy}
+      - MEGA_SERVICE_HOST_IP=${MEGA_SERVICE_HOST_IP}
+      - EMBEDDING_SERVICE_HOST_IP=${EMBEDDING_SERVICE_HOST_IP}
+      - RETRIEVER_SERVICE_HOST_IP=${RETRIEVER_SERVICE_HOST_IP}
+      - RERANK_SERVICE_HOST_IP=${RERANK_SERVICE_HOST_IP}
+      - LLM_SERVICE_HOST_IP=${LLM_SERVICE_HOST_IP}
+    ipc: host
+    restart: always
+  chaqna-xeon-ui-server:
+    image: opea/chatqna-ui:latest
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:169]: same update here.

File: ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+      - no_proxy=${no_proxy}
+      - https_proxy=${https_proxy}
+      - http_proxy=${http_proxy}
+      - CHAT_BASE_URL=${BACKEND_SERVICE_ENDPOINT}
+      - UPLOAD_FILE_BASE_URL=${DATAPREP_SERVICE_ENDPOINT}
+      - GET_FILE=${DATAPREP_GET_FILE_ENDPOINT}
+      - DELETE_FILE=${DATAPREP_DELETE_FILE_ENDPOINT}
+    ipc: host
+    restart: always
+  chaqna-xeon-conversation-ui-server:
```
Comment (chensuyue) [inline ChatQnA/docker_compose/intel/cpu/xeon/compose_pinecone.yaml:185]: You can remove this server.

File: ChatQnA/tests/test_compose_pinecone_on_xeon.sh [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+    echo "In stop docker"
+    echo $WORKPATH
+    cd $WORKPATH/docker_compose/intel/cpu/xeon/
+    docker compose -f compose_pinecone.yaml down
+}
+
+function main() {
+
+    stop_docker
+    if [[ "$IMAGE_REPO" == "" ]]; then build_docker_images; fi
```
Comment (chensuyue) [inline ChatQnA/tests/test_compose_pinecone_on_xeon.sh:261]: if [[ "$IMAGE_REPO" == "opea" ]]; then build_docker_images; fi

File: ChatQnA/tests/test_compose_pinecone_on_xeon.sh [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+    export PINECONE_INDEX_NAME="langchain-test"
+    export INDEX_NAME="langchain-test"
+    export HUGGINGFACEHUB_API_TOKEN=${HUGGINGFACEHUB_API_TOKEN}
+
+    # Start Docker Containers
+    docker compose -f compose_pinecone.yaml up -d > ${LOG_PATH}/start_services_with_compose.log
+
+    n=0
+    until [[ "$n" -ge 200 ]]; do
+        docker logs tgi-service > tgi_service_start.log
```
Comment (chensuyue) [inline ChatQnA/tests/test_compose_pinecone_on_xeon.sh:48]: Can you write this log into `${LOG_PATH}/tgi_service_start.log`, so we can see what happened.

**Review by chensuyue — APPROVED** [review]

**Review by ZePan110 — APPROVED** [review]

**Pull request merged**